### ISSUE TITLE
External scheduler can get out of sync when hosts are added if the host shows up and is added to the external scheduler while containers are in flux

### ISSUE THREAD

**Issue by cjellick** [rancher/rancher#8639]
**Rancher Versions:**
Server: 1.6.0-rc3
healthcheck:
ipsec:
network-services:
scheduler:
kubernetes (if applicable):

**Docker Version:**
N/A

**OS and where are the hosts located? (cloud, bare metal, etc):**
N/A

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node rancher

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
@chrisurwin reported this while working on his rebalancer project. The logs from his scheduler container are attached.

1. Setup a rancher cluster with three nodes
2. **magic bad race condition**
3. Launch a service and incrementally scale it up to 9

**Results:** 
Expected: Containers in service should balance out, relative to all containers on the hosts across the three hosts
Actual: One host if favored more than the others

Note: this is an intermittent race condition due to the scheduler not starting with the correct number of containers.

[sched.logs.zip](https://github.com/rancher/rancher/files/971998/sched.logs.zip)

**Comment by cjellick** [comment]
I don't believe this is new in 1.6, Chris just happened to hit it with a 1.6 rc

**Comment by cjellick** [comment]
I've attach logs from @sangeethah's setup that experience the same problem. Its clear that we have a problem initializing hosts. It seams the problem is that 
Two things we want to do to fix this:
1. Make initialization process smart so that it doesnt have a gap between populating from metadata and receiving events from cattle.
2. Make a resync process that runs on an interval that effectively pauses (locks) the scheduler, rebuilds from metadata, and then unpauses

Regarding fix #1:
I think in fresh environments where images have to be pulled, the scheduler can take up to a minute to get to healthy (please validate if this is true). But, populating hosts from metadata happens immediately at startup. So, what happens is that we populate a host from MD and it has zero containers, scheduler is unhealthy, so the infra containers going to that host do not go through the external scheduler. Then scheduler gets healthy and starts receiving events, but at this point, the host's count is off by the number of containers it missed initially.

To fix this, we should wait to initialize from metadata until we receive the first scheudler event from cattle. At that point, we know cattle thinks the scheduler is ready-to-go and that we wont miss any containers

Regarding fix #2:
We need to lock the entire scheduler, resync with metatadta, unlock the entire scheduler. Interval should be configurable but can be every 3-5 mintues

For both fixes, we need to make sure the metadata sync process is fast. This may involve rewriting the logic to be more efficient.

**Comment by cjellick** [comment]
[san-sched.logs.zip](https://github.com/rancher/rancher/files/974299/san-sched.logs.zip)

**Comment by deniseschannon** [comment]
@StrongMonkey Can you look at #8179 for reasons why it might be happening that was reported by @janeczku?

**Comment by StrongMonkey** [comment]
https://github.com/rancher/scheduler/pull/47 is implemented based on your design here. @cjellick

**Comment by StrongMonkey** [comment]
Here are the steps to reproduce it:
1. Have a rancher setup with scheduler enabled. Stop the health-check service, then restart scheduler container to make sure it goes to initializing state. 
2. Add a new host. Waiting for all the infra service to come up. Then start the health-check service, and wait scheduler to become green. Then launch a service with several containers. You will see all the containers will land on the new host, breaking the balance rule.
To prove the new scheduler image works, just follow the exact same steps above. The correct behavior would be that the service containers will try to balance itself across the hosts.

**Comment by cjellick** [comment]
Another way to (possible) reproduce, which is closer to what is actually being reported:
1. Get rancher images staged on two hosts. You can do this by just adding the hosts to a rancher cluster and then removing them.
2. Prepare a third host so that you can add it to rancher manually.
  * To be clear, at this point you should have zero hosts in rancher, but three VMs ready to add to rancher. Two of those should have all the rancher images staged on them
3. On all three hosts, get the docker run command ready so that all you have to do is hit enter.
4. On the two hosts with staged images, hit enter on the docker run command to add them to rancher.
5. Flip over to the UI and when you see infrastructure containers being added to the first two hosts, jump back to the third host and hit enter to add it to rancher.

This will very likely induce a race condition where the scheduler container is initializaing while the third container is just beginning to add its infrastructure containers. Because they arent yet in metadata and the scheduler container is not yet receiving events (it only receives events when it is healthy), the infrastructure containers will likely get miscounted for the third host. You can observe this in the scheduler's container logs by search for `Adding resource pool [instanceReservation]`. A correct run for that will have either 8 or 9 in this part of the message: `with total 1000000 and used 8 for host`. A bad run will have a smaller number like 0 or 1.

For the fix, you will still see the third host being added with a small number like `Adding resource pool [instanceReservation] with total 1000000 and used 1 for host`, but when you ask rancher to schedule the next container by lauching a service (or launching a container from the container page*), the scheduler will resync and you should all the hosts added with the correct number of instanceReservations.

 * I said specifically "from the container page" because adding a container to a specific host on the Hosts page does not kick off the resync logic.

**Comment by deniseschannon** [comment]
Available in v1.6.1-rc9 with scheduler:v0.8.0

**Comment by galal-hussein** [comment]
Rancher server: v1.6.1-rc9

I followed the steps that @cjellick provided, after add the third host i found that the instancereservation number is low as expected:
```
time="2017-06-01T19:58:00Z" level=info msg="Adding resource pool [instanceReservation] with total 1000000 and used 0 for host  2a555389-1c3a-4641-a5fe-1f8d8305cec2"
```
And after adding a container (from the Container page), i can see that all hosts has the correct number of instanceReservations:
```
time="2017-06-01T20:00:47Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 9 for host  9ac47d7c-af76-4d68-ab41-d287680f6f66" 
time="2017-06-01T20:00:47Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 8 for host  2a555389-1c3a-4641-a5fe-1f8d8305cec2" 
time="2017-06-01T20:00:47Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 8 for host  c273ec36-5065-4e85-89c2-45dc71e9517f" 
time="2017-06-01T20:02:46Z" level=info msg="Reserving [{Resource:instanceReservation Amount:1}] for 2a555389-1c3a-4641-a5fe-1f8d8305cec2. Force=true" 
time="2017-06-01T20:02:46Z" level=info msg="New pool amount on host 2a555389-1c3a-4641-a5fe-1f8d8305cec2: instanceReservation total: 1000000 used: 9." 
```
